cmd/nv2/gencert.go

```diff
@@ -0,0 +1,209 @@
+}
+
+var generateCertCommand = &cli.Command{
+	Name:      "generate",
+	Usage:     "Generates a test crt and key file.",
+	ArgsUsage: "[host]",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:     "rsaBits",
+			Usage:    "--rsaBits 3072",
+			Required: false,
+			Value:    3072,
+		},
+		&cli.StringFlag{
+			Name:     "not-after",
+			Usage:    "--not-after 2006-01-02T15:04:05-07:00 (default is 1 year)",
+			Required: false,
+		},
+	},
+
+	Action: runGenerateCert,
+}
+
+func ensureKeysDir() (string, error) {
+
+	// Expected to ensure ~/.notation/keys
+	dirname, err := os.UserHomeDir()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	keysDir := filepath.Join(dirname, KEYS_BASE_DIR, "keys")
+	fsStat, err := os.Stat(keysDir)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(keysDir, 0700) // Only user has permissions on this directory
+		if err != nil {
+			return "", err
+		}
+	} else if fsStat.IsDir() == false {
+		return "", fmt.Errorf("%s should be a directory", keysDir)
+	}
+
+	return keysDir, nil
+}
+
+//ref: https://golang.org/src/crypto/tls/generate_cert.go
+
+func runGenerateCert(ctx *cli.Context) error {
+
+	host := ctx.Args().First()
+	if len(host) == 0 {
+		return errors.New("Missing required [host] parameter")
+	}
+
+	// Set certificate validity
+	notBefore := time.Now()
+	notAfter := notBefore.Add(time.Duration(365 * 24 * time.Hour))
+
+	expiry := ctx.String("not-after")
+	var err error
+	if len(expiry) != 0 {
+		notAfter, err = time.Parse(time.RFC3339, expiry)
+		if err != nil {
+			return fmt.Errorf("Invalid --not-after %s value specified %v", expiry, err)
+		}
+	}
+
+	if notAfter.Before(notBefore) {
+		return fmt.Errorf("Invalid --not-after that is earlier than not-before [%s] specified", notBefore.Format(time.RFC3339))
+	}
+
+	// Generate RSA Bits
+	rsaBits := ctx.Int("rsaBits")
+	var priv crypto.Signer
+
+	fmt.Printf("Generating RSA Key with %d bits\n", rsaBits)
+	priv, err = rsa.GenerateKey(rand.Reader, rsaBits)
+
+	if err != nil {
+		return fmt.Errorf("Failed to generate private key: %v", err)
+	}
+
+	// ECDSA, ED25519 and RSA subject keys should have the DigitalSignature
+	// KeyUsage bits set in the x509.Certificate template
+	keyUsage := x509.KeyUsageDigitalSignature
+	extKeyUsage := []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning}
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return fmt.Errorf("Failed to generate serial number: %v", err)
+	}
+
+	template := x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           extKeyUsage,
+		BasicConstraintsValid: true,
+	}
+
+	hosts := strings.Split(host, ",")
+	for _, h := range hosts {
+		if ip := net.ParseIP(h); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, h)
+		}
+	}
```

Comment:
as this is codesigning not ssl certificate, we don't need to support DNSName or ip-address.

Reply:
We still need them as the `x509` package checks `SAN` instead of `CN` starting from golang `1.15`. See https://golang.org/doc/go1.15#commonname